### PR TITLE
Generalized CycleFold

### PR DESCRIPTION
**Pull request by winderica**
This PR depends on #95.

In Nova and HyperNova, the challenge `r` has 128 bits. On the other hand, `r` is sampled from the prime field in Protogalaxy and therefore may be longer than 128 bits.

This PR refactors CycleFold to support both cases. Now, in Nova and HyperNova, we create impls (e.g., `NovaCycleFoldConfig`) for `CycleFoldConfig` and set `CycleFoldConfig::RANDOMNESS_BIT_LENGTH = NOVA_N_BITS_RO`. While in (the incoming implementation of) Protogalaxy, we can set `CycleFoldConfig::RANDOMNESS_BIT_LENGTH = F::MODULUS_BIT_SIZE`.

In addition, now `CommittedInstance` and `Witness` from Nova are re-exported as `CycleFoldCommittedInstance` and `CycleFoldWitness` in `cyclefold.rs`, so the CycleFold instances and witnesses in HyperNova and ProtoGalaxy would have more clear type names.

### CODE REVIEW

**Comment by arnaucube** [comment]
Hi @winderica , I'm very sorry because in this PR you've been suffering from git conflicts with `main` branch with the last merged PRs.
I think that in the next days there will not be much changes on the `main` branch that affect this PR changes, so if it works for you, you can rebase to the latest `main` version, and we prioritize reviewing this PR to avoid postponing it more (and making you keep rebasing and git-conflict-fixing).

**Comment by winderica** [comment]
No worries @arnaucube ! I just rebased this PR onto the current `main` 😄

**Review by arnaucube — APPROVED** [review]
Nice, the changes make a lot of sense! LGTM!

File: folding-schemes/src/folding/hypernova/circuits.rs [hunk 1]
```diff
@@ -861,20 +874,9 @@ where
...
         // Fold cf1_u_i & cf_U_i into cf1_U_{i+1}
-        let cf_U_i1 = NIFSFullGadget::<C2, GC2>::fold_committed_instance(
-            cf_r_bits,
-            cf_r_nonnat,
-            cf_cmT,
-            cf_U_i,
-            cf_u_i,
-        )?;
+        let cf_U_i1 =
+            NIFSFullGadget::<C2, GC2>::fold_committed_instance(cf_r_bits, cf_cmT, cf_U_i, cf_u_i)?;
```
Comment (CPerezz) [inline folding-schemes/src/folding/hypernova/circuits.rs:879]: I like this! Much more readable!

**Review by CPerezz — APPROVED** [review]
LGTM! Awesome work!

**Review by dmpierre — APPROVED** [review]
Nice work :)

**Pull request merged**